src/ext/coms.c

```diff
@@ -0,0 +1,440 @@
+    char *data = stack->data;
+    size_t size = stack->size;
+
+    memset(stack, 0, sizeof(ddtrace_coms_stack_t));
+    memset(data, 0, size);
+
+    stack->data = data;
+    stack->size = size;
+}
+
+void gc_stacks() {
+    for (int i = 0; i < DD_TRACE_COMS_STACKS_BACKLOG_SIZE; i++) {
+        ddtrace_coms_stack_t *stack = ddtrace_coms_global_state.stacks[i];
+
+        if (stack) {
+            if (ddtrace_coms_is_stack_unused(stack) && atomic_load(&stack->bytes_written) == 0) {
+                ddtrace_coms_global_state.stacks[i] = NULL;
+                free(stack);
+            } else {
+                stack->gc_cycles_count++;
+            }
+        }
+    }
+}
+
+static void init() {
+    ddtrace_coms_stack_t *stack = new_stack();
+    if (!ddtrace_coms_global_state.stacks) {
+        ddtrace_coms_global_state.stacks = calloc(DD_TRACE_COMS_STACKS_BACKLOG_SIZE, sizeof(ddtrace_coms_stack_t *));
+    }
+
+    atomic_store(&ddtrace_coms_global_state.next_group_id, 1);
+    atomic_store(&ddtrace_coms_global_state.current_stack, stack);
+}
+
+uint32_t ddtrace_coms_rotate_stack() {
+    ddtrace_coms_stack_t *stack = NULL;
+    ddtrace_coms_stack_t *current_stack = atomic_load(&ddtrace_coms_global_state.current_stack);
+
+    for (int i = 0; i < DD_TRACE_COMS_STACKS_BACKLOG_SIZE; i++) {
+        ddtrace_coms_stack_t *stack_tmp = ddtrace_coms_global_state.stacks[i];
+        if (stack_tmp) {
+            if (atomic_load(&stack_tmp->refcount) == 0 && atomic_load(&stack_tmp->bytes_written) == 0) {
+                stack = stack_tmp;
+                recycle_stack(stack_tmp);
+                ddtrace_coms_global_state.stacks[i] = current_stack;
+                current_stack = NULL;
+                break;
+            }
+        }
+    }
+
+    // attempt to freeup stack storage
+    gc_stacks();
+
+    if (current_stack != NULL) {
+        for (int i = 0; i < DD_TRACE_COMS_STACKS_BACKLOG_SIZE; i++) {
+            if (!ddtrace_coms_global_state.stacks[i]) {
+                ddtrace_coms_global_state.stacks[i] = current_stack;
+                current_stack = NULL;
+            }
+        }
+    }
+
+    // old current stack was stored so set a new stack
+    if (current_stack == NULL) {
+        if (!stack) {
+            stack = new_stack();
+        }
+
+        atomic_store(&ddtrace_coms_global_state.current_stack, stack);
+        return 0;
+    }
+    // if we couldn't store new stack i tem
+    return ENOMEM;
+}
+
+uint32_t ddtrace_coms_flush_data(uint32_t group_id, const char *data, size_t size) {
+    if (!data) {
+        return 0;
+    }
+
+    if (data && size == 0) {
+        size = strlen(data);
+    }
+
+    if (size == 0) {
+        return 0;
+    }
+
+    if (store_data(group_id, data, size) == 0) {
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
+group_id_t ddtrace_coms_next_group_id() { return atomic_fetch_add(&ddtrace_coms_global_state.next_group_id, 1); }
+
+uint32_t ddtrace_coms_initialize() {
```

Comment:
Is there a reason for wrapping `init()` here instead of just renaming `init()` to `ddtrace_coms_initialize()`? :)

Reply:
Yeah, this is "legacy code" now :) thanks for spotting that! In the beginning I had an Idea more things would jump in there but nope :)